lib/spark.js

```diff
@@ -58,8 +64,30 @@ export default class Spark {
             stats.firstByteAt = new Date()
           }
           stats.byteLength += value.byteLength
+
+          // We want to limit how large content we are willing to download.
+          // 1. To make sure we don't spend too much time (and network bandwidth) on a single task,
+          //    so that we can complete more tasks per round
+          // 2. Until we have streaming hashes, we need to keep the entire payload in memory, and so
+          //    we need to put an upper limit on how much memory we consume.
+          if (stats.byteLength > MAX_CAR_SIZE) {
+            stats.carTooLarge = true
+            break
```

Comment:
Discussion point: The currently proposed implementation will set `stats.success` to `true`. Do we want to set it to `false` to indicate a failure? I can see arguments for both versions.

- The SP provided all the data we asked for. It's not their fault that we stopped fetching more data. From this point of view, the retrieval was successful.
- We don't know whether the SP would serve us the entire CAR payload or whether it would stop responding later (after the 200MB mark). Smart SPs can learn about this size limit in SPARK and configure their HTTP endpoints to do exactly what I described - serve the first 200MB only. From the point of measuring the reliability of retrievals, we don't know if the given CID can be retrieved in full, so we should not claim the task was a success.

Personally, I am leaning towards flagging this case as a successful retrieval. We can use the extra flag `carTooLarge` to monitor how many CIDs are too large to be retrieved by SPARK and then figure out a mitigation strategy.

Reply:
I've been having issues with `.success`, as a success can mean different things depending on how you look at it. (as this case shows again). I'd propose to make the data we produce here unopinionated, and let the data consumer decide.

For example, from an SP perspective, you wouldn't want to call this a successful retrieval, as we don't know whether everything could have been retrieved. On the other hand, from a Spark perspective, we can say that what it tried to do worked, it only couldn't finish it.

This problem also shows in the grafana dashboard definition:

```sql
  COUNT(1) FILTER (WHERE success IS TRUE AND status_code >= 200 AND status_code < 300) as "Success",
```

I think we should rename both the field and the semantics. I'll try to come up with a PR, or let's both of us discuss.

Reply:
https://github.com/filecoin-station/spark/pull/33